File: bin/main/io/signalbird/sdk/Storage.kt
package io.signalbird.sdk

import org.json.JSONObject
import java.util.concurrent.locks.ReentrantLock
import kotlin.concurrent.withLock

/**
 * Ziyaretçi kimliğinin saklandığı yer.
 *
 * Android'de `SharedPreferences` sarmalayın. Sır cihazda kalmazsa kullanıcı
 * uygulamayı her açtığında sohbet geçmişini kaybeder — bu yüzden saklama
 * zorunludur, isteğe bağlı değil.
 */
interface SignalbirdStorage {
    fun get(key: String): String?
    fun set(key: String, value: String)
    fun remove(key: String)
}

/** Depo verilmezse: yalnız süreç ömrü boyunca yaşayan bellek deposu. */
class MemoryStorage : SignalbirdStorage {
    private val map = mutableMapOf<String, String>()

    override fun get(key: String): String? = map[key]
    override fun set(key: String, value: String) { map[key] = value }
    override fun remove(key: String) { map.remove(key) }
}

/**
 * Ziyaretçi kimliğinin kilitli kutusu.
 *
 * Sır iki yerden okunur (istek başlıkları ve `currentVisitorId`) ve oturum
 * açılışında yazılır; ikisi farklı iş parçacığında olabilir.
 */
internal class VisitorStore(
    private val storage: SignalbirdStorage,
    private val appKey: String,
) {
    private val lock = ReentrantLock()
    private var cached: Pair<String, String>? = null
    private var loaded = false

    val id: String? get() = current()?.first
    val secret: String? get() = current()?.second

    fun save(id: String, secret: String) = lock.withLock {
        cached = id to secret
        loaded = true

        storage.set(
            STORAGE_KEY,
            JSONObject(mapOf("id" to id, "secret" to secret, "appKey" to appKey)).toString(),
        )
    }

    fun clear() = lock.withLock {
        cached = null
        loaded = true
        storage.remove(STORAGE_KEY)
    }

    /**
     * Anahtar değiştiyse (uygulama döndürüldü, farklı ortam) kimlik geçersizdir:
     * eski sırla yapılan her çağrı 401 alırdı ve sohbet sessizce ölürdü.
     */
    private fun current(): Pair<String, String>? = lock.withLock {
        if (loaded) return cached

        loaded = true

        val text = storage.get(STORAGE_KEY) ?: return null

        cached = try {
            val parsed = JSONObject(text)
            val id = parsed.optString("id")
            val secret = parsed.optString("secret")

            if (id.isEmpty() || secret.isEmpty() || parsed.optString("appKey") != appKey) {
                null
            } else {
                id to secret
            }
        } catch (_: Exception) {
            null
        }

        cached
    }

    private companion object {
        const val STORAGE_KEY = "sb_visitor"
    }
}

File: src/swift/Sources/Signalbird/Storage.swift
import Foundation

/// Ziyaretçi kimliğinin saklandığı yer.
///
/// Varsayılan `UserDefaults`'tur. Sır cihazda kalmazsa kullanıcı uygulamayı
/// her açtığında sohbet geçmişini kaybeder — bu yüzden saklama zorunludur,
/// isteğe bağlı değil. Keychain isteyen kendi uyarlamasını verir.
public protocol SignalbirdStorage: Sendable {
    func get(_ key: String) -> String?
    func set(_ key: String, _ value: String)
    func remove(_ key: String)
}

/// `UserDefaults` `Sendable` işaretli değildir ama okuma/yazması iş parçacığı
/// güvenlidir (Apple belgeleri). `@unchecked` bunu açıkça üstlenir; alternatifi
/// her erişimi bir kuyruğa taşımaktı — üç satırlık bir depo için fazla.
public struct UserDefaultsStorage: SignalbirdStorage, @unchecked Sendable {
    private let defaults: UserDefaults

    public init(defaults: UserDefaults = .standard) {
        self.defaults = defaults
    }

    public func get(_ key: String) -> String? { defaults.string(forKey: key) }
    public func set(_ key: String, _ value: String) { defaults.set(value, forKey: key) }
    public func remove(_ key: String) { defaults.removeObject(forKey: key) }
}

/// Ziyaretçi kimliğinin kilitli kutusu.
///
/// Sır iki yerden okunur — istek başlıkları ve `currentVisitorID` — ve oturum
/// açılışında yazılır. Küçük ama paylaşılan bir durum olduğu için kilidi
/// kendisi taşır; istemcinin tamamını eşzamanlılık kurallarına hapsetmeye
/// gerek kalmaz.
final class VisitorStore: @unchecked Sendable {
    private static let storageKey = "sb_visitor"

    private let storage: SignalbirdStorage
    private let publicKey: String
    private let lock = NSLock()
    private var cached: (id: String, secret: String)?
    private var loaded = false

    init(storage: SignalbirdStorage, publicKey: String) {
        self.storage = storage
        self.publicKey = publicKey
    }

    var id: String? { current()?.id }

    var secret: String? { current()?.secret }

    func save(id: String, secret: String) {
        lock.lock()
        defer { lock.unlock() }

        cached = (id, secret)
        loaded = true

        let payload: [String: Any] = ["id": id, "secret": secret, "publicKey": publicKey]

        if let raw = try? JSONSerialization.data(withJSONObject: payload),
           let text = String(data: raw, encoding: .utf8) {
            storage.set(Self.storageKey, text)
        }
    }

    func clear() {
        lock.lock()
        defer { lock.unlock() }

        cached = nil
        loaded = true
        storage.remove(Self.storageKey)
    }

    /// Anahtar değiştiyse (uygulama döndürüldü, farklı ortam) kimlik geçersizdir:
    /// eski sırla yapılan her çağrı 401 alırdı ve sohbet sessizce ölürdü.
    private func current() -> (id: String, secret: String)? {
        lock.lock()
        defer { lock.unlock() }

        if loaded { return cached }

        loaded = true

        guard let text = storage.get(Self.storageKey),
              let raw = text.data(using: .utf8),
              let parsed = try? JSONSerialization.jsonObject(with: raw) as? [String: Any],
              let id = parsed["id"] as? String,
              let secret = parsed["secret"] as? String,
              parsed["publicKey"] as? String == publicKey
        else {
            return nil
        }

        cached = (id, secret)

        return cached
    }
}
